### PR TITLE
Refactor iterable

### LINKED ISSUE #776: Refactor Iterator
The iterators in the Core library could be better. Some ideas:

- Rename them to Iterate instead of Iterable (nobody will assume a mutable state in current Candy)
- Return `Empty` or `More [item, rest]` instead of a result. Sure, this will make us lose `result.map` and such, but having a few well-placed pattern matches makes the code much more readable.
- Be more accurate with responsibilities. `is iter = function.is0 iter` is not sufficient

### PR DESCRIPTION
**Pull request by MarcelGarus**
Closes: #776 

### Checklist

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: packages/Random/rngs/pseudo/xorshift.candy [hunk 1]
```diff
@@ -27,17 +27,17 @@ is generator := generator %
...
     | lowEndianBytesToU32
   }
 
   state = ifElse
-    seed | iterable.fromList | iterable.all { it -> it | equals 0 }
-    { [X: 0x0BAD5EED, Y: 0x0BAD5EED, Z: 0x0BAD5EED, W: 0x0BAD5EED] }
-    { [X: getU32 0, Y: getU32 1, Z: getU32 2, W: getU32 3] }
+      seed | iterator.fromList | iterator.all { it -> it | equals 0 }
+      { [X: 0x0BAD5EED, Y: 0x0BAD5EED, Z: 0x0BAD5EED, W: 0x0BAD5EED] }
+      { [X: getU32 0, Y: getU32 1, Z: getU32 2, W: getU32 3] }
```
Comment (JonasWanke) [inline packages/Random/rngs/pseudo/xorshift.candy]:
```suggestion
    seed | iterator.fromList | iterator.all { it -> it | equals 0 }
    { [X: 0x0BAD5EED, Y: 0x0BAD5EED, Z: 0x0BAD5EED, W: 0x0BAD5EED] }
    { [X: getU32 0, Y: getU32 1, Z: getU32 2, W: getU32 3] }
```

File: packages/Core/iterator.candy [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+  generateWithState [iterator, Index: 0] { state ->
+    [iterator, index] = state
+    iterator | next %
+      Empty -> Empty
+      More [item, rest] -> [Item: [item, index], State: [Iterator: rest, Index: index | int.add 1]]
+  }
+
+ensure iterator guard :=
+  # Useful for ensuring that a `needs` holds for each item of the iterator.
+  function.is1 guard
```
Comment (JonasWanke) [inline packages/Core/iterator.candy]:
```suggestion
  needs (function.is1 guard)
```

File: packages/Core/iterator.candy [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+    items = iterator | take size | toList
+    length = items | list.length
+    isLongEnough = ifElse allowPartialWindows { length | int.isPositive } { length | equals size }
+    ifElse isLongEnough { Ok [Item: items, State: iterator | skip step] } { Error Empty }
+  }
+chunked iterator size :=
+  needs (int.is size)
+  needs (int.isPositive size)
+
+  iterator | wrapSafe { needs False "`chunked` needs an iterator." } | windowed size size True
```
Comment (JonasWanke) [inline packages/Core/iterator.candy:269]: The tests got deleted

File: packages/Core/iterator.candy [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+wrapSafe iterator notIteratorGuard =
+  needs (function.is0 notIteratorGuard)
+  panicUsingGuard = {
+    function.run notIteratorGuard
+    needs False "The guard needs to panic."
+  }
+
+  nextFunction = iterator %
+    Iterator function -> function
+    _ -> panicUsingGuard
```
Comment (JonasWanke) [inline packages/Core/iterator.candy]:
```suggestion
    _ -> function.run panicUsingGuard
```

**Review by JonasWanke — CHANGES_REQUESTED** [review]

**Comment by MarcelGarus** [comment]
Some notes for changes that don't fit the PR title:

- I fixed a use-after-free introduced by #745 
- I erroneously accepted https://github.com/candy-lang/candy/pull/744#discussion_r1373202603, causing failed asserts of `!self.is_negative()`

**Comment by MarcelGarus** [comment]
Surprisingly, `wrapSafe` doesn't seem to impact performance too badly and it results in a much better dev experience when errors occur. That's why I'm inclined to leave it in for now.

File: compiler/vm/src/heap/object_inline/int.rs [hunk 5]
```diff
@@ -92,7 +92,7 @@ impl InlineInt {
                     // One 1 is necessary for the sign.
                     i64::BITS - lhs.leading_ones() + 1
                 } else {
-                    lhs.bit_length()
```
Comment (JonasWanke) [inline compiler/vm/src/heap/object_inline/int.rs]: What was the problem here? This is only called on non-negative numbers

File: packages/Core/iterator.candy [hunk 6]
```diff
@@ -69,6 +71,7 @@ generateWithState initialState mapper :=
       mapper state %
         More [item, state] -> More [item, Rest: recurse state]
         Empty -> Empty
+        a -> needs False "The `mapper` is bad. It should return Empty or More [item, state], but it returned this: {a}"
```
Comment (JonasWanke) [inline packages/Core/iterator.candy]:
```suggestion
        a -> needs False "The `mapper` is bad. It should return `Empty` or `More [item, state]`, but it returned this: `{a}`"
```

**Review by JonasWanke — APPROVED** [review]

**Pull request merged**